### ISSUE TITLE
Autofill support

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#13015]
We should make sure our text fields trigger the autofill support:
   https://developer.android.com/guide/topics/text/autofill.html

**Comment by acidjazz** [comment]
Doesn't this framework work whenever an input is recognized natively as an input? 
Why wouldn't they via flutter?

**Comment by orospakr** [comment]
fwiw, I have the following form:

```dart
          Form(
              child: Column(
                children: <Widget>[
                  TextFormField(
                    keyboardType: TextInputType.emailAddress,
                    decoration: new InputDecoration(
                      hintText: "Email",
                      labelText: "Email"
                    ),
                  ),
                  TextFormField(
                    obscureText: true,
                    decoration: new InputDecoration(
                      labelText: "Password"
                    ),
                  )
                ],
              )
            ),
```

And 1Password doesn't seem to be triggering.

**Comment by jonahwilliams** [comment]
The Android auto-fill framework wasn't added until SDK 26, so I doubt we support it currently since Flutter's target SDK is still 22

**Comment by Hixie** [comment]
We will definitely have to support it. We will just have to call back to nothing on older Androids.

**Comment by jonahwilliams** [comment]
re-reading what I wrote it sounds a little negative ...

Anyway, we definitely don't support it currently - but we absolutely can. I am updating our android sdk version this week, which will enable the use of this API on newer devices

**Comment by cbracken** [comment]
Similarly we should ensure that we support the same for the password field changes in iOS 11 (may already work?) and the upcoming changes in iOS 12.

**Comment by jonahwilliams** [comment]
Investigated updating the android sdk and buildroot this week, found that it's not really feasible unless we are willing to invest time into refactoring buildroot.  see https://github.com/flutter/flutter/issues/17780 for more details - this is blocking implementing autofill

**Comment by Hixie** [comment]
Might be a dupe of https://github.com/flutter/flutter/issues/14047

**Comment by focux** [comment]
Any news about this?

**Comment by w4mxl** [comment]
Any new progress? 🧐

**Comment by stuartsoft** [comment]
Any updates?

**Comment by stuartsoft** [comment]
Ok, so I have confirmation from @dnfield  that we should now be unblocked. What's the current status on this ticket? What needs to be done? Can I help?

**Comment by dnfield** [comment]
https://developer.android.com/guide/topics/text/autofill-optimize#virtual needs to be implemented in the Android TextInputPlugin.

**Comment by dnfield** [comment]
Or perhaps https://developer.android.com/guide/topics/text/autofill-optimize#callback

**Comment by stuartsoft** [comment]
Well, I am pleased to see that lastpass is at least able to acknowledge the presence of email/password fields and prompts to autofill. Unfortunately that's where the good new stops, because it can't seem to actually perform the autofill action :(

**Comment by gincos** [comment]
Any news on this??

**Comment by markqq** [comment]
Any news on this?

**Comment by shinayser** [comment]
Any news on this?

**Comment by saharh** [comment]
Any news on this?

**Comment by tinusg** [comment]
Would love an update on this. Right now the login procedure is the main thing that makes my app feel non-native.

**Comment by saharh** [comment]
Hi guys,

Tired of waiting so I've created a flutter plugin:
https://pub.dev/packages/flutter_autofill

Usage is simple, just wrap Autofill widget around your Textfield as follows:
`          Autofill(
                onAutofilled: (val) {
                   // set value in controller & cursor position after auto-filled value
                  _emailController.value = TextEditingValue(text: val, selection: TextSelection.fromPosition(TextPosition(offset: val.length)));
                },
                autofillHints: [FlutterAutofill.AUTOFILL_HINT_EMAIL_ADDRESS],
                autofillType: FlutterAutofill.AUTOFILL_TYPE_TEXT,
                textController: _emailController,
                child: TextField(
                  controller: _emailController,
                  decoration: InputDecoration(hintText: "Please enter your email"),
                ),
              ),
`

Works using Android autofill [virtual structure support](https://developer.android.com/guide/topics/text/autofill-optimize#virtual)

Would appreciate comments/improvements!

**Comment by emmet63001** [comment]
hi guys, cant find how to connect google chrome and safari autofill saved passwords and login to flutter app, who has any idea ?

**Comment by hacker1024** [comment]
> Hi guys,
> 
> Tired of waiting so I've created a flutter plugin:
> https://pub.dev/packages/flutter_autofill
> 
> Usage is simple, just wrap Autofill widget around your Textfield as follows:
> 
> ```
>           Autofill(
>                 onAutofilled: (val) {
>                    // set value in controller & cursor position after auto-filled value
>                   _emailController.value = TextEditingValue(text: val, selection: TextSelection.fromPosition(TextPosition(offset: val.length)));
>                 },
>                 autofillHints: [FlutterAutofill.AUTOFILL_HINT_EMAIL_ADDRESS],
>                 autofillType: FlutterAutofill.AUTOFILL_TYPE_TEXT,
>                 textController: _emailController,
>                 child: TextField(
>                   controller: _emailController,
>                   decoration: InputDecoration(hintText: "Please enter your email"),
>                 ),
>               ),
> ```
> 
> Works using Android autofill [virtual structure support](https://developer.android.com/guide/topics/text/autofill-optimize#virtual)
> 
> Would appreciate comments/improvements!

This is great, but a major issue is that there's no autofill option when long-pressing a text field. This needs to be built-in to Flutter.

**Comment by HansMuller** [comment]
I added the cujpertino and material labels because I assume that developers will expect integration of autofill with the TextField's text selection menu (per https://github.com/flutter/flutter/issues/13015#issuecomment-536173301)

**Comment by oth-libs** [comment]
Any news on this on flutter side?

**Comment by oth-libs** [comment]
was anyone able to successfully integrate the solution from @saharh? I have a few issues, I put the same code as in the sample but unable to save or restore the email/password

**Comment by justinmc** [comment]
@genert was asking about helping out with the effort to get Autofill working.  It's going to be a decently large amount of work to do this right.  Here's the work that I see needing to be done for this:

  * Agree on a proposed solution
    - Understand how autofill works natively on Android and iOS now.
    - Come up with an engine API that ideally covers both in a generic way.
    - Come up with an API for TextField and CupetinoTextField that ties into the engine.
    - Make sure that this solution would also work for non-text fields like CupertinoDatePicker, etc., if related.
    - Make sure that a Flutter dev could tie into this API with their own custom form widgets as well.
  * Implement the engine API.
  * Tie the engine API into TextField/TextFormField/CupertinoTextField.
  * Tie the engine API into other form widgets like CupertinoTextField.

Right now I think we need to start out by doing some research and agreeing on a potential solution to this.

**Comment by AlexVincent525** [comment]
As an implement with engine https://github.com/flutter/engine/pull/13878 , does it finished the engine side?

**Comment by justinmc** [comment]
Looks like that could be what we need for iOS in the engine.  I just gave a quick review.

**Comment by kaarens93** [comment]
Hi, any updates to this??

**Comment by shinayser** [comment]
This feature has 165 👍 + 7 ❤, isn't that enough for flutter team to take a look? 😢

**Comment by LongCatIsLooong** [comment]
We're working on it and hope to release something in the first quarter of 2020.

**Comment by dnfield** [comment]
I have hidden a bunch of comments in this thread that were mainly "any updates" comments.

Please avoid commenting "same here", "+1", or "any updates" - it makes it harder to find relevant content in the issue, pings all subscribers, and does not help surface to the team the priority of the issue.

Instead, thumbs up the original comment. If you wish to subscribe to updates, click the subscribe button in the right hand column of the issue. Comments should be reserved for information pertinent to the issue such as what progress you're making on finishing the feature, request for design review, updates in platform API docs to consider, blockers that need to be resolved, etc.

**Comment by justinmc** [comment]
@jcivelli-google and anyone else involved in this effort (@LongCatIsLooong, @GaryQian, @jason-simmons, @dkwingsmt), here is a quick summary of all the info we have about this problem and next steps.  If I'm missing anything please let me know.

## Outstanding work

  1. We'll need to write up a design doc about how exactly this will work and get buy-in for the approach, including a way to support both iOS and Android with minimal friction.
  1. Tie in to both iOS and Android's autofill APIs in the engine.
  1. Wire up the engine code to all form fields in the framework in a way that exposes the necessary features of both platforms.

### Tying into Android autofill
@jcivelli-google posted an issue about this: https://github.com/flutter/flutter/issues/49385. See the  [Android docs on virtual structure](https://developer.android.com/guide/topics/text/autofill-optimize#virtual) for how we can inform Android of our widget structure.  Also the [Android docs on HintConstants](https://developer.android.com/reference/androidx/autofill/HintConstants) for how to inform Android of various input types.

### Tying in iOS autofill
There is also an outstanding PR for engine iOS autofill support that appears to be abandoned, but may have some useful ideas (https://github.com/flutter/engine/pull/13878).  Either way, we'll need to see how closely compatible iOS's autofill is with Android's.  iOS uses [UITextContentType](https://developer.apple.com/documentation/uikit/uitextcontenttype) to identify field types (similar to HintConstants, linked above, for Android).

**Comment by jonahwilliams** [comment]
> 2. Tie in to both iOS and Android's autofill APIs in the engine.

Don't forget the web!

**Comment by justinmc** [comment]
Also bringing up this iOS password autofill issue: https://github.com/flutter/flutter/issues/32285

**Comment by ganttastic** [comment]
I just pushed a couple of draft PRs I've been working on for autofill. They are still WIP, but I wanted to get them out there for comment. I got excited and dove into the engine, so let me know if I'm jumping ahead too quickly.

Engine: https://github.com/flutter/engine/pull/16227
Framework: https://github.com/flutter/flutter/pull/49762

* iOS is functionally complete, but I'm having an issue where auto-fill suggestions only appear if not using the default text keyboard. If I use email or alphanumeric style keyboards, it works like a charm.
* Android still has some work to be done. If anyone has any guidance on if I'm on the right track there, it would be appreciated.
* Web needs to be started. (and I don't know where to start there)

I plan on diving into it a bit more this weekend. I can remove the Android implementation if needed to just push the iOS functionality. Let me know if you have any questions or concerns.

**Comment by giaur500** [comment]
What is current status on Android? I guess it works partially, or not yet? I don't know what guidance you need actually, because on native Android it works by default, I don't need to do anything to enable built in Google's autofill. Only put email/password text views and autofill works without any additional effort from my side

**Comment by ganttastic** [comment]
@giaur500 The logic to flow from the dart layer to the flutter engine is roughly in place. Because these are not native android input views, we're having to rely on the lower level frameworks. The guidance I'm needing is more on the modifications needed to flutter engine itself. @LongCatIsLooong could already be working on the engine side for this.

**Comment by yjbanov** [comment]
@Hixie @LongCatIsLooong I added the web label as this is also important on the web and requires engine-side work. Also, would it be possible to move this to the near-term goals milestone? Users really really want this.

**Comment by fmatosqg** [comment]
From the comments I feel like smart lock is out of scope.  https://developer.android.com/distribute/best-practices/develop/smart-lock-for-passwords

**Comment by soorur** [comment]
+1

**Comment by jdeltoft** [comment]
Any idea when this might be scheduled for?

**Comment by ajaygautam** [comment]
This would be great to have. Saves the user from typing over and over again!

**Comment by LongCatIsLooong** [comment]
This should be working on the master branch now. Please open new issues for feature requests / bug reports!

### Known issues: 
- https://github.com/flutter/flutter/issues/41508: on iOS we don't currently support [`UIKeyboardType.namePhonePad`](https://developer.apple.com/documentation/uikit/uikeyboardtype/namephonepad) keyboard type, which seems to be a requirement for triggering name/telephone/address autofill on iOS.
- AutofillHint `newPassword` mapping is currently missing in the iOS text input plugin.

### Upcoming features:
- Allow the system autofill service to save user input: https://github.com/flutter/flutter/issues/55613
- Add more platform-specific autofill flags, for example `passwordRules` on iOS.

**Comment by Vofchuk** [comment]
I have this issu also on the web version... any idea how to fix?

**Comment by Freundschaft** [comment]
as far as I can see, the latest master channel already implements this for flutter web, I can see the autocomplete attributes being set, but apparently the input text fields are generated on the fly, so chrome continues to offer only saving for the password input field, anyone else who can reproduce this?

**Comment by nturgut** [comment]
@Freundschaft Thanks for the question. We are working on that part, there needs to be changes on the framework, later we will merge the Flutter Web changes. The tracking issue: https://github.com/flutter/flutter/issues/55613

**Comment by Levi-Lesches** [comment]
Chrome has offered to save and autofill my password on a website I built with Flutter Web. I believe I was on the master channel when I compiled it. 

> chrome continues to offer only saving for the password input field

Yes, I have the same issue, Chrome only offers to save and fill the password but not the username.

**Comment by nturgut** [comment]
@Levi-Lesches yes we didn't merge the ["save" solution](https://github.com/flutter/flutter/issues/55613) yet. Since some fields handled different by the browsers you can see the save popups for some fields.

**Comment by max202021** [comment]
@nturgut Is the "save" solution merged? Is it tracked using this issue or there is some other issue?

**Comment by jacob-hutchings** [comment]
Wondering if anyone else has this autofill problem.

On iOS, when I have multiple text fields on a screen with different autofill hints, tapping between the text fields does not update the autofill correctly. Sometimes the autofill just doesn't show up. Sometimes the time the first autofill item just stays and never switches. Sometimes when I leave the app and come back, all the autofills have change to "Passwords" with the key icon.

![IMG_1790](https://user-images.githubusercontent.com/3528029/90426649-a5c6ed00-e07e-11ea-9149-4b1ac51a7328.PNG)

![IMG_1789](https://user-images.githubusercontent.com/3528029/90426657-a95a7400-e07e-11ea-9ae3-6524a5fc8a7c.PNG)


Might be related to https://github.com/flutter/flutter/issues/59374

**Comment by stx** [comment]
> Wondering if anyone else has this autofill problem.
> 
> On iOS, when I have multiple text fields on a screen with different autofill hints, tapping between the text fields does not update the autofill correctly. Sometimes the autofill just doesn't show up. Sometimes the first autofill item just stays and never switches. Sometimes when I leave the app and come back, all the autofills have change to "Passwords" with the key icon.
> 
> ![IMG_1790](https://user-images.githubusercontent.com/3528029/90426649-a5c6ed00-e07e-11ea-9149-4b1ac51a7328.PNG)
> 
> ![IMG_1789](https://user-images.githubusercontent.com/3528029/90426657-a95a7400-e07e-11ea-9ae3-6524a5fc8a7c.PNG)
> 
> Might be related to #59374

Yep, same here.

**Comment by nturgut** [comment]
> @nturgut Is the "save" solution merged? Is it tracked using this issue or there is some other issue?

Save solution is merged. Sorry I just saw this cc.

**Comment by LongCatIsLooong** [comment]
@jacob-hutchings @stx thanks for reporting this issue! What's your flutter version? From the screenshot it seems it happens on non-password fields right? For those fields I think we reuse the same ios view and I've never had this issue, but it's possible ios autofill caches the content type internally. Could you provide a minimal repro for investigation?

**Comment by jacob-hutchings** [comment]
> @jacob-hutchings @stx thanks for reporting this issue! What's your flutter version? From the screenshot it seems it happens on non-password fields right? For those fields I think we reuse the same ios view and I've never had this issue, but it's possible ios autofill caches the content type internally. Could you provide a minimal repro for investigation?

Thank you for looking at this @LongCatIsLooong !

Here is my Flutter version info:

Flutter (Channel stable, 1.20.2, on Mac OS X 10.15.4 19E287, locale en-US)

I tested the following code on my physical iPhone 8 and also the iOS simulator iPhone SE (2nd Gen). Autofill straight up never works with the following code for me:

```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Autofill repro',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController _firstNameController = TextEditingController();
  TextEditingController _lastNameController = TextEditingController();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Autofill repro'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(18.0),
        child: Column(
          children: <Widget>[
            TextField(
              controller: _firstNameController,
              decoration: InputDecoration(
                hintText: 'Given Name',
              ),
              autofillHints: [AutofillHints.givenName],
            ),
            TextField(
              controller: _lastNameController,
              decoration: InputDecoration(
                hintText: 'Family Name',
              ),
              autofillHints: [AutofillHints.familyName],
            ),
          ],
        ),
      ),
    );
  }
}

```

I repeatedly tested this, rebuilding, force quitting and launching, and it never showed.

After adding `autocorrect: False,` to the text fields and testing a bunch more times:
- I could sometimes get autofill to show but not very often. 
- If I type in some text and backspace it all the way, autofill shows up pretty regularly, however, after it shows up, switching between the text fields does not update the autofill:

![IMG_3ADABE2A7E45-1](https://user-images.githubusercontent.com/3528029/90435459-48d23380-e08c-11ea-9750-f8bcc6cf0df5.jpeg)

After testing tens of times I finally got it to update one time, but I'm not sure how I got it to work. 

The behavior is very inconsistent. 

Please let me know if there is additional info you need.

**Comment by giaur500** [comment]
My login form is defined as follows:

```
Form buildLoginForm() {
    final citiesListControl = buildCitiesSelector();

    final email = TextFormField(
      enabled: _isCitySelected,
      keyboardType: TextInputType.emailAddress,
      key: Key('email'),
      autofocus: false,
      initialValue: '',
      decoration: InputDecoration(
        hintText: 'E-mail',
        contentPadding: EdgeInsets.fromLTRB(20.0, 10.0, 20.0, 10.0),
        border: OutlineInputBorder(borderRadius: BorderRadius.circular(32.0)),
        focusedBorder: OutlineInputBorder(
          borderSide: const BorderSide(color: Colors.black26, width: 2.0),
          borderRadius: BorderRadius.circular(32.0),
        ),
      ),
      onSaved: (value) => _email = value,
      validator: (val) => !EmailValidator.Validate(val, true)
          ? Strings.msg_email_invalid
          : null,
        autofillHints: [AutofillHints.email]
    );

    final password = TextFormField(
      enabled: _isCitySelected,
      key: Key('password'),
      autofocus: false,
      initialValue: '',
      obscureText: true,
      decoration: InputDecoration(
        hintText: Strings.password_hint,
        contentPadding: EdgeInsets.fromLTRB(20.0, 10.0, 20.0, 10.0),
        border: OutlineInputBorder(borderRadius: BorderRadius.circular(32.0)),
        focusedBorder: OutlineInputBorder(
          borderSide: const BorderSide(color: Colors.black26, width: 2.0),
          borderRadius: BorderRadius.circular(32.0),
        ),
      ),
      onSaved: (value) => _password = value,
      validator: (val) =>
          val.length < 4 ? Strings.msg_password_too_short : null,
        autofillHints: [AutofillHints.password]
    );

    final loginButton = Padding(
      padding: EdgeInsets.symmetric(vertical: 2.0),
      child: RaisedButton(
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(24.0),
        ),
        onPressed: _isCitySelected
            ? () {
                if (_loginFormKey.currentState.validate()) {
                  _loginFormKey.currentState.save();
                  _loginEmailPassword();
                }
              }
            : null,
        padding: EdgeInsets.all(12.0),
        color: Colors.lightBlueAccent,
        child: Text(Strings.btn_login, style: TextStyle(color: Colors.white)),
      ),
    );

    final guestLoginButton = Padding(
      padding: EdgeInsets.symmetric(vertical: 0.0),
      child: RaisedButton(
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(24.0),
        ),
        onPressed: _isCitySelected
            ? () {
                _loginFormKey.currentState.save();
                _loginGuest();
              }
            : null,
        padding: EdgeInsets.all(12.0),
        color: Colors.green,
        child: Text(Strings.btn_login_guest,
            style: TextStyle(color: Colors.white)),
      ),
    );

    final statusLabel = Padding(
        padding: EdgeInsets.symmetric(vertical: 16.0),
        child: Text(_status ?? "", style: TextStyle(color: Colors.red)));

    return Form(
      key: this._loginFormKey,
      child: Scaffold(
        backgroundColor: Colors.white,
        body: Center(
          child: ListView(
            shrinkWrap: true,
            padding: EdgeInsets.only(left: 24.0, right: 24.0),
            children: <Widget>[
              citiesListControl,
              SizedBox(height: 32.0),
              email,
              SizedBox(height: 8.0),
              password,
              SizedBox(height: 24.0),
              loginButton,
              guestLoginButton,
              statusLabel
            ],
          ),
        ),
      ),
    );
  }
```
On Android, I am never getting any autofill. Is that anything wrong in my code? Issue is closed, so I guess functionality is finished, so what's wrong?

**Comment by bierbaumtim** [comment]
Every post I saw about Autofill they wrap the Widget were the TextInputFields in with an AutofillGroup. I does it the same way and it works fine.
So maybe that would solve the problems.
Also for iOS you need to define an Runner.entitlements file with associated domains and webcredentials to be able to use them.

**Comment by mbaleczny** [comment]
Wrapping with `AutofillGroup` seems to do the magic. Also, in order to be able to see "save data" pop-up I had to use `Form` and save it on submit / successful login.
Tested on Android with autofill services: 
* Google - save data worked, update password didn't worked.
* Samsung Pass - none worked.

Have you managed to make it working with other password managers?

**Comment by abc873693** [comment]
> Wrapping with `AutofillGroup` seems to do the magic. Also, in order to be able to see "save data" pop-up I had to use `Form` and save it on submit / successful login.
> Tested on Android with autofill services:
> 
> * Google - save data worked, update password didn't worked.
> * Samsung Pass - none worked.
> 
> Have you managed to make it working with other password managers?

Ref this issue https://github.com/flutter/flutter/issues/55613#event-3635992635.
This feature not release stable channel.

**Comment by giaur500** [comment]
I tried to wrap `Form` or `ListView` with `AutofillGroup` but still no autofill shown.

**Comment by bochigarcia** [comment]
Hello, it works !!! Wrap the fields with AutoFillGroup and add to each field your autoFillHint is magic in web. The autocomplete with email and password started to work. Thanks

**Comment by bazmichael** [comment]
> H

could you please share some piece of that logic?

**Comment by lukepighetti** [comment]
`AutofillGroup` is the magic
